File: src/tricked/mcts/features.py
"""
Standard Documentation for features.py.

This module supplies the core execution logic for the `mcts` namespace, heavily typed and tested for production distribution.
"""

import torch
from tricked_engine import GameStateExt as GameState

from tricked.env.constants import TOTAL_TRIANGLES, get_neighbors
from tricked.env.pieces import get_piece_overlay, get_valid_placement_mask


def extract_feature(
    state: GameState, 
    history: list[int] | None = None, 
    action_history: list[int] | None = None, 
    difficulty: int = 1
) -> torch.Tensor:
    
    feature = torch.zeros(20, TOTAL_TRIANGLES, dtype=torch.float32)

    def fill_channel(channel_idx: int, board_int: int) -> None:
        for i in range(TOTAL_TRIANGLES):
            if (board_int >> i) & 1:
                feature[channel_idx, i] = 1.0

    history = history or []
    
    fill_channel(0, state.board)
    
    for i in range(1, 8):
        if len(history) >= i:
            fill_channel(i, history[-i])
        else:
            fill_channel(i, state.board)  

    action_history = action_history or []
    for i in range(3):
        if len(action_history) > i:
            act = action_history[-(i+1)]
            slot = act // TOTAL_TRIANGLES
            idx = act % TOTAL_TRIANGLES
            if idx < TOTAL_TRIANGLES:
                feature[8 + i, idx] = (slot + 1) * 0.33

    for slot in range(3):
        p_id = state.available[slot]
        if p_id == -1:
            continue
        overlay = get_piece_overlay(p_id)
        valid_mask = get_valid_placement_mask(p_id, state.board)
        for i in range(TOTAL_TRIANGLES):
            if overlay[i] == 1:
                feature[11 + (slot * 2), i] = 1.0
            if valid_mask[i] == 1:
                feature[12 + (slot * 2), i] = 1.0

    feature[17, :] = 1.0 / 22.0

    feature[18, :] = float(difficulty) / 6.0  

    for i in range(TOTAL_TRIANGLES):
        is_filled = (state.board >> i) & 1
        if not is_filled:
            
            surrounded = True
            for n in get_neighbors(i):
                if not ((state.board >> n) & 1):
                    surrounded = False
                    break
            if surrounded:
                feature[19, i] = 1.0  

    return feature

File: src/features.rs
use crate::board::GameStateExt;
use crate::constants::STANDARD_PIECES;
use crate::neighbors::NEIGHBOR_MASKS;

const TOTAL_TRIANGLES: usize = 96;

pub fn extract_feature_native(
    state: &GameStateExt,
    history: Option<Vec<u128>>,
    action_history: Option<Vec<i32>>,
    difficulty: i32,
) -> Vec<f32> {
    let mut feature = vec![0.0_f32; 20 * TOTAL_TRIANGLES];

    let history = history.unwrap_or_default();
    let action_history = action_history.unwrap_or_default();

    let fill_channel = |feat: &mut [f32], channel_idx: usize, board_int: u128| {
        let offset = channel_idx * TOTAL_TRIANGLES;
        for i in 0..96 {
            if (board_int >> i) & 1 == 1 {
                feat[offset + i] = 1.0;
            }
        }
    };

    fill_channel(&mut feature, 0, state.board);

    // channel 1-7: history
    for i in 1..=7 {
        if history.len() >= i {
            fill_channel(&mut feature, i, history[history.len() - i]);
        } else {
            fill_channel(&mut feature, i, state.board);
        }
    }

    // channel 8-10: action history
    for i in 0..3 {
        if action_history.len() > i {
            let act = action_history[action_history.len() - (i + 1)];
            let slot = act / (TOTAL_TRIANGLES as i32);
            let idx = (act % (TOTAL_TRIANGLES as i32)) as usize;
            if idx < TOTAL_TRIANGLES {
                feature[(8 + i) * TOTAL_TRIANGLES + idx] = (slot as f32 + 1.0) * 0.33;
            }
        }
    }

    // channel 11-16: pieces overlay and valid mask
    for slot in 0..3 {
        let p_id = state.available[slot];
        if p_id == -1 {
            continue;
        }
        let p_idx = p_id as usize;
        let mut overlay = [0_u8; 96];
        let mut valid_mask = [0_u8; 96];

        for m in &STANDARD_PIECES[p_idx] {
            if *m == 0 {
                continue;
            }
            // overlay
            for i in 0..96 {
                if (*m & (1 << i)) != 0 {
                    overlay[i] = 1;
                }
            }
            // valid mask
            if (state.board & *m) == 0 {
                for i in 0..96 {
                    if (*m & (1 << i)) != 0 {
                        valid_mask[i] = 1;
                    }
                }
            }
        }

        for i in 0..TOTAL_TRIANGLES {
            if overlay[i] == 1 {
                feature[(11 + slot * 2) * TOTAL_TRIANGLES + i] = 1.0;
            }
            if valid_mask[i] == 1 {
                feature[(12 + slot * 2) * TOTAL_TRIANGLES + i] = 1.0;
            }
        }
    }

    // channel 17: empty
    for i in 0..TOTAL_TRIANGLES {
        feature[17 * TOTAL_TRIANGLES + i] = 1.0 / 22.0;
    }

    // channel 18: difficulty
    let diff_val = difficulty as f32 / 6.0;
    for i in 0..TOTAL_TRIANGLES {
        feature[18 * TOTAL_TRIANGLES + i] = diff_val;
    }

    // channel 19: hole checking
    for i in 0..TOTAL_TRIANGLES {
        let is_filled = (state.board >> i) & 1 == 1;
        if !is_filled {
            let neighbors = NEIGHBOR_MASKS[i];
            if (state.board & neighbors) == neighbors {
                feature[19 * TOTAL_TRIANGLES + i] = 1.0;
            }
        }
    }

    feature
}


